File: examples/crypto/bn254/g1-g2-basics.ts
import { BN254 } from '../../../src/crypto/bn254/BN254.js';

/**
 * BN254 G1/G2 Basic Operations
 *
 * Demonstrates fundamental elliptic curve operations:
 * - Generator points
 * - Point addition
 * - Scalar multiplication
 * - Point equality
 * - Point serialization/deserialization
 * - Subgroup membership checks
 */

console.log('=== BN254 G1/G2 Basic Operations ===\n');

// 1. G1 Generator and Basic Operations
console.log('1. G1 Point Operations');
console.log('-'.repeat(40));

// Get G1 generator
const g1Gen = BN254.G1.generator();
console.log('G1 generator created');
console.log(`Is zero: ${BN254.G1.isZero(g1Gen)}`);
console.log(`Is on curve: ${BN254.G1.isOnCurve(g1Gen)}\n`);

// Scalar multiplication
const scalar1 = 5n;
const g1Point1 = BN254.G1.mul(g1Gen, scalar1);
console.log(`G1 generator × ${scalar1}:`);
console.log(`Result is on curve: ${BN254.G1.isOnCurve(g1Point1)}\n`);

// Point addition
const scalar2 = 3n;
const g1Point2 = BN254.G1.mul(g1Gen, scalar2);
const g1Sum = BN254.G1.add(g1Point1, g1Point2);
console.log(`(G1 × ${scalar1}) + (G1 × ${scalar2}):`);
console.log(`Result is on curve: ${BN254.G1.isOnCurve(g1Sum)}`);

// Verify linearity: (5 + 3) * G1 should equal (5 * G1) + (3 * G1)
const g1Direct = BN254.G1.mul(g1Gen, scalar1 + scalar2);
const linearityHolds = BN254.G1.equal(g1Sum, g1Direct);
console.log(`Linearity check (5+3)×G1 = (5×G1)+(3×G1): ${linearityHolds}\n`);

// Point doubling
const g1Doubled = BN254.G1.double(g1Gen);
const g1Times2 = BN254.G1.mul(g1Gen, 2n);
const doublingCorrect = BN254.G1.equal(g1Doubled, g1Times2);
console.log(`Point doubling: 2×G1 = G1+G1: ${doublingCorrect}\n`);

// 2. G2 Operations (Extension Field)
console.log('2. G2 Point Operations (Fp2)');
console.log('-'.repeat(40));

// Get G2 generator
const g2Gen = BN254.G2.generator();
console.log('G2 generator created');
console.log(`Is zero: ${BN254.G2.isZero(g2Gen)}`);
console.log(`Is on curve: ${BN254.G2.isOnCurve(g2Gen)}`);
console.log(`Is in subgroup: ${BN254.G2.isInSubgroup(g2Gen)}\n`);

// Scalar multiplication
const scalar3 = 7n;
const g2Point1 = BN254.G2.mul(g2Gen, scalar3);
console.log(`G2 generator × ${scalar3}:`);
console.log(`Result is on curve: ${BN254.G2.isOnCurve(g2Point1)}`);
console.log(`Result is in subgroup: ${BN254.G2.isInSubgroup(g2Point1)}\n`);

// Point addition
const scalar4 = 11n;
const g2Point2 = BN254.G2.mul(g2Gen, scalar4);
const g2Sum = BN254.G2.add(g2Point1, g2Point2);
console.log(`(G2 × ${scalar3}) + (G2 × ${scalar4}):`);
console.log(`Result is on curve: ${BN254.G2.isOnCurve(g2Sum)}`);
console.log(`Result is in subgroup: ${BN254.G2.isInSubgroup(g2Sum)}\n`);

// Verify G2 linearity
const g2Direct = BN254.G2.mul(g2Gen, scalar3 + scalar4);
const g2LinearityHolds = BN254.G2.equal(g2Sum, g2Direct);
console.log(`G2 Linearity check (7+11)×G2 = (7×G2)+(11×G2): ${g2LinearityHolds}\n`);

// 3. Point Negation
console.log('3. Point Negation');
console.log('-'.repeat(40));

const g1Neg = BN254.G1.negate(g1Gen);
const g1ZeroFromAdd = BN254.G1.add(g1Gen, g1Neg);
console.log(`G1 + (-G1) = infinity: ${BN254.G1.isZero(g1ZeroFromAdd)}`);

const g2Neg = BN254.G2.negate(g2Gen);
const g2ZeroFromAdd = BN254.G2.add(g2Gen, g2Neg);
console.log(`G2 + (-G2) = infinity: ${BN254.G2.isZero(g2ZeroFromAdd)}\n`);

// 4. Infinity Point
console.log('4. Infinity Point (Group Identity)');
console.log('-'.repeat(40));

const g1Infinity = BN254.G1.infinity();
console.log(`G1 infinity is zero: ${BN254.G1.isZero(g1Infinity)}`);

// Adding infinity is identity
const g1PlusInfinity = BN254.G1.add(g1Point1, g1Infinity);
const identityHolds = BN254.G1.equal(g1Point1, g1PlusInfinity);
console.log(`P + infinity = P: ${identityHolds}`);

const g2Infinity = BN254.G2.infinity();
console.log(`G2 infinity is zero: ${BN254.G2.isZero(g2Infinity)}\n`);

// 5. Serialization (for precompiles and storage)
console.log('5. Point Serialization');
console.log('-'.repeat(40));

// Serialize G1 point (64 bytes: 32 for x, 32 for y)
const g1Serialized = BN254.serializeG1(g1Point1);
console.log(`G1 serialized length: ${g1Serialized.length} bytes`);
console.log(`First 8 bytes (x): ${Array.from(g1Serialized.slice(0, 8)).map(b => b.toString(16).padStart(2, '0')).join('')}...`);

// Deserialize back
const g1Deserialized = BN254.deserializeG1(g1Serialized);
const g1RoundTrip = BN254.G1.equal(g1Point1, g1Deserialized);
console.log(`Round-trip successful: ${g1RoundTrip}\n`);

// Serialize G2 point (128 bytes: 32 each for x.c0, x.c1, y.c0, y.c1)
const g2Serialized = BN254.serializeG2(g2Point1);
console.log(`G2 serialized length: ${g2Serialized.length} bytes`);
console.log(`First 8 bytes (x.c0): ${Array.from(g2Serialized.slice(0, 8)).map(b => b.toString(16).padStart(2, '0')).join('')}...`);

// Deserialize back
const g2Deserialized = BN254.deserializeG2(g2Serialized);
const g2RoundTrip = BN254.G2.equal(g2Point1, g2Deserialized);
console.log(`Round-trip successful: ${g2RoundTrip}\n`);

// 6. Large Scalar Multiplication
console.log('6. Large Scalar Multiplication');
console.log('-'.repeat(40));

const largeScalar = 123456789012345678901234567890n;
const g1Large = BN254.G1.mul(g1Gen, largeScalar);
console.log(`G1 × ${largeScalar}:`);
console.log(`Result is on curve: ${BN254.G1.isOnCurve(g1Large)}`);

// Demonstrate that order matters: n*G + m*G = (n+m)*G
const n = 999999n;
const m = 111111n;
const nG = BN254.G1.mul(g1Gen, n);
const mG = BN254.G1.mul(g1Gen, m);
const sum = BN254.G1.add(nG, mG);
const direct = BN254.G1.mul(g1Gen, n + m);
console.log(`Distributive property: ${n}×G + ${m}×G = ${n + m}×G: ${BN254.G1.equal(sum, direct)}\n`);

console.log('=== Complete ===');

File: examples/crypto/bn254/g1-g2-basics.zig
const std = @import("std");
const crypto = @import("crypto");

const G1 = crypto.bn254.G1;
const G2 = crypto.bn254.G2;
const Fr = crypto.bn254.Fr;
const FpMont = crypto.bn254.FpMont;
const Fp2Mont = crypto.bn254.Fp2Mont;

/// BN254 G1/G2 Basic Operations
///
/// Demonstrates fundamental elliptic curve operations:
/// - Generator points
/// - Point addition
/// - Scalar multiplication
/// - Point equality
/// - Subgroup membership checks
pub fn main() !void {
    const stdout = std.io.getStdOut().writer();

    try stdout.print("=== BN254 G1/G2 Basic Operations ===\n\n", .{});

    // 1. G1 Generator and Basic Operations
    try stdout.print("1. G1 Point Operations\n", .{});
    try stdout.print("{s}\n", .{"-" ** 40});

    // Get G1 generator (hardcoded values from curve_parameters.zig)
    const g1_gen = G1{
        .x = FpMont.init(1),
        .y = FpMont.init(2),
        .z = FpMont.init(1),
    };

    try stdout.print("G1 generator created\n", .{});
    try stdout.print("Is infinity: {}\n", .{g1_gen.isInfinity()});
    try stdout.print("Is on curve: {}\n\n", .{g1_gen.isOnCurve()});

    // Scalar multiplication
    const scalar1 = Fr.init(5);
    const g1_point1 = try g1_gen.mul(&scalar1);
    try stdout.print("G1 generator × 5:\n", .{});
    try stdout.print("Result is on curve: {}\n\n", .{g1_point1.isOnCurve()});

    // Point addition
    const scalar2 = Fr.init(3);
    const g1_point2 = try g1_gen.mul(&scalar2);
    const g1_sum = g1_point1.add(&g1_point2);
    try stdout.print("(G1 × 5) + (G1 × 3):\n", .{});
    try stdout.print("Result is on curve: {}\n", .{g1_sum.isOnCurve()});

    // Verify linearity: (5 + 3) * G1 should equal (5 * G1) + (3 * G1)
    const scalar_sum = scalar1.add(&scalar2);
    const g1_direct = try g1_gen.mul(&scalar_sum);
    const linearity_holds = g1_sum.equal(&g1_direct);
    try stdout.print("Linearity check (5+3)×G1 = (5×G1)+(3×G1): {}\n\n", .{linearity_holds});

    // Point doubling
    const g1_doubled = g1_gen.double();
    const scalar_two = Fr.init(2);
    const g1_times2 = try g1_gen.mul(&scalar_two);
    const doubling_correct = g1_doubled.equal(&g1_times2);
    try stdout.print("Point doubling: 2×G1 = G1+G1: {}\n\n", .{doubling_correct});

    // 2. G2 Operations (Extension Field)
    try stdout.print("2. G2 Point Operations (Fp2)\n", .{});
    try stdout.print("{s}\n", .{"-" ** 40});

    // Get G2 generator (hardcoded values from curve_parameters.zig)
    const g2_gen = G2{
        .x = Fp2Mont.initFromInt(
            0x1800deef121f1e76426a00665e5c4479674322d4f75edadd46debd5cd992f6ed,
            0x198e9393920d483a7260bfb731fb5d25f1aa493335a9e71297e485b7aef312c2,
        ),
        .y = Fp2Mont.initFromInt(
            0x12c85ea5db8c6deb4aab71808dcb408fe3d1e7690c43d37b4ce6cc0166fa7daa,
            0x090689d0585ff075ec9e99ad690c3395bc4b313370b38ef355acdadcd122975b,
        ),
        .z = Fp2Mont.initFromInt(1, 0),
    };

    try stdout.print("G2 generator created\n", .{});
    try stdout.print("Is infinity: {}\n", .{g2_gen.isInfinity()});
    try stdout.print("Is on curve: {}\n", .{g2_gen.isOnCurve()});
    try stdout.print("Is in subgroup: {}\n\n", .{g2_gen.isInSubgroup()});

    // Scalar multiplication
    const scalar3 = Fr.init(7);
    const g2_point1 = g2_gen.mul(&scalar3);
    try stdout.print("G2 generator × 7:\n", .{});
    try stdout.print("Result is on curve: {}\n", .{g2_point1.isOnCurve()});
    try stdout.print("Result is in subgroup: {}\n\n", .{g2_point1.isInSubgroup()});

    // Point addition
    const scalar4 = Fr.init(11);
    const g2_point2 = g2_gen.mul(&scalar4);
    const g2_sum = g2_point1.add(&g2_point2);
    try stdout.print("(G2 × 7) + (G2 × 11):\n", .{});
    try stdout.print("Result is on curve: {}\n", .{g2_sum.isOnCurve()});
    try stdout.print("Result is in subgroup: {}\n\n", .{g2_sum.isInSubgroup()});

    // Verify G2 linearity
    const scalar_sum2 = scalar3.add(&scalar4);
    const g2_direct = g2_gen.mul(&scalar_sum2);
    const g2_linearity_holds = g2_sum.equal(&g2_direct);
    try stdout.print("G2 Linearity check (7+11)×G2 = (7×G2)+(11×G2): {}\n\n", .{g2_linearity_holds});

    // 3. Point Negation
    try stdout.print("3. Point Negation\n", .{});
    try stdout.print("{s}\n", .{"-" ** 40});

    const g1_neg = g1_gen.neg();
    const g1_zero_from_add = g1_gen.add(&g1_neg);
    try stdout.print("G1 + (-G1) = infinity: {}\n", .{g1_zero_from_add.isInfinity()});

    const g2_neg = g2_gen.neg();
    const g2_zero_from_add = g2_gen.add(&g2_neg);
    try stdout.print("G2 + (-G2) = infinity: {}\n\n", .{g2_zero_from_add.isInfinity()});

    // 4. Infinity Point
    try stdout.print("4. Infinity Point (Group Identity)\n", .{});
    try stdout.print("{s}\n", .{"-" ** 40});

    const g1_infinity = G1{
        .x = FpMont.init(0),
        .y = FpMont.init(0),
        .z = FpMont.init(0),
    };
    try stdout.print("G1 infinity is zero: {}\n", .{g1_infinity.isInfinity()});

    // Adding infinity is identity
    const g1_plus_infinity = g1_point1.add(&g1_infinity);
    const identity_holds = g1_point1.equal(&g1_plus_infinity);
    try stdout.print("P + infinity = P: {}\n", .{identity_holds});

    const g2_infinity = G2{
        .x = Fp2Mont.initFromInt(0, 0),
        .y = Fp2Mont.initFromInt(0, 0),
        .z = Fp2Mont.initFromInt(0, 0),
    };
    try stdout.print("G2 infinity is zero: {}\n\n", .{g2_infinity.isInfinity()});

    // 5. Large Scalar Multiplication
    try stdout.print("5. Large Scalar Multiplication\n", .{});
    try stdout.print("{s}\n", .{"-" ** 40});

    const large_scalar = Fr.init(123456789);
    const g1_large = try g1_gen.mul(&large_scalar);
    try stdout.print("G1 × 123456789:\n", .{});
    try stdout.print("Result is on curve: {}\n", .{g1_large.isOnCurve()});

    // Demonstrate that order matters: n*G + m*G = (n+m)*G
    const n = Fr.init(999999);
    const m = Fr.init(111111);
    const nG = try g1_gen.mul(&n);
    const mG = try g1_gen.mul(&m);
    const sum = nG.add(&mG);
    const n_plus_m = n.add(&m);
    const direct = try g1_gen.mul(&n_plus_m);
    try stdout.print("Distributive property: 999999×G + 111111×G = 1111110×G: {}\n\n", .{sum.equal(&direct)});

    try stdout.print("=== Complete ===\n", .{});
}
